Support CDATA in writeStringTextNode

MelkorCC opened the pull request:
Closes #10127, Based on code and discussions in #15638 

Summary:
Functionality:
This change affects how the `writeStringTextNode()` function in `ol/format/xsd.js` handles data containing whitespaces (\n\t, multiple spaces, spaces at the start and the end of the text) and special XML characters (<'>&"). Strings containing these will now be wrapped in CDATA section(s). In order to determine if a string is affected by this a new local method `checkCData` was introduced that takes a string and checks if it should be wrapped in CDATA section(s). Checking the string has a worse case performance of O(N), the string will only be iterated once.
The algorithm that breaks a string safely into (potentially) multiple CDATA sections is based on code written in #15638.

As discussed in #15638 `ol/format/KML.js` was also adapted, calls to `writeCDATASection` can now safely be replaced by calls to `writeStringTextNode`.

Testing:
In order to support proper testing changes needed to be done to the `test-extensions`, in particular `getChildNodes` had to be adapted to consider CDATA sections as well as `assertElementNodesEqual` for the actual comparison. Unlike in the text node comparison no whitespaces are removed, the "raw" `nodeValue` is directly compared.

Only one existing test had to be adapted, a KML test contained an unnecessary CDATA section.

New tests were added in `ol/format/xsd.test.js`. These tests cover the expected behavior for "normal" strings and data, strings containing whitespaces, strings containing special characters and strings containing data leading to multiple CDATA sections.

As mentioned in #15638 the chromium implementation of DOMParser has currently a bug and can incorrectly merge two CDATA sections when parsing a XML string. Because of this tests resulting in multiple CDATA sections construct the XMLDocument programmatically, all other tests utilize a XML string containing the expected XML and parsing for comparison.
This bug does only affect testing, the implemented functionality is not affected. If this bug is fixed in future chromium releases we can modify these tests, the expected XML string is provided alongside the test in a comment.


Open Questions:

- [x] In theory it is possible to pass data that is NOT a string to `writeStringTextNode`. Since one KML test did this with a number i supported this with tests for the primitive types boolean and number. In theory one could also pass objects to `writeStringTextNode`, this use case is currently not supported. If this IS a use case we want to support we should probably call `toString` on all input going into `writeStringTextNode`.
- [x] I am unsure if all the special XML characters are really problematic for textnodes. Of course < and & are problematic, but i don't see how '>" are problematic. I added these for now, but if somebody that is more knowledgeable with XML and special cases confirms that these are not problematic in text nodes we can drop them again.

ahocevar reviewed (changes requested):
Thanks for your effort on this, @MelkorCC!

I added a few simplification suggestions that will probably also improve performance, see in-code comments.

> In theory it is possible to pass data that is NOT a string to writeStringTextNode. Since one KML test did this with a number i supported this with tests for the primitive types boolean and number. In theory one could also pass objects to writeStringTextNode, this use case is currently not supported. If this IS a use case we want to support we should probably call toString on all input going into writeStringTextNode.

I don't find a call site in KML.js where `writeStringTextNode()` is called with a `number`. No need to test anything but strings, we have type checks to avoid incorrect types being passed to `writeStringTextNode`, and `toString()` will be called internally if something different than a string is passed to a node's `textContent`. Also, for boolean values, there is a `writeBooleanTextNode()` function.

> I am unsure if all the special XML characters are really problematic for textnodes. Of course < and & are problematic, but i don't see how '>" are problematic. I added these for now, but if somebody that is more knowledgeable with XML and special cases confirms that these are not problematic in text nodes we can drop them again.

In a text node (which we're dealing with here), only `"&"` and `"<"` need to be escaped. The regular expression in my code suggestions is already adjusted to only contain these two.

Regarding tests: I don't see a reason to use `createElementNS('http://www.w3.org/1999/xhtml', 'text')` for creating test nodes - I think there isn't even a `text` element available in XHTML. That said,`document.createElement('text')` is preferred.

> In order to support proper testing changes needed to be done to the test-extensions, in particular getChildNodes had to be adapted to consider CDATA sections as well as assertElementNodesEqual for the actual comparison. Unlike in the text node comparison no whitespaces are removed, the "raw" nodeValue is directly compared.

I don't think that's necessary. Instead of using `xmleq()`, I'd suggest using `node.textContent` to compare the `writeStringTextNode()` node with the reference node.

MelkorCC commented:
@ahocevar thank you very much for your quick feedback

> I don't find a call site in KML.js where writeStringTextNode() is called with a number.

My bad, maybe it was boolean. But in my very first prototype i didn't have the `typeof === string` check in `writeStringTextNode` which did result in an error since a string function was called on something that didn't have such a function. I think i traced it back to KML and [this ](https://github.com/openlayers/openlayers/blob/main/test/browser/spec/ol/format/kml.test.js#L1903) testcase, but i am not sure right now. I am not a KML expert so i didn't look into it. The only thing i can say is that i definitly had cases where something not being a string was passed to `writeStringTextNode`. Hence my testcoverage and my questions regarding non-primitive stuff being potentially passed to it.


> (...) , and toString() will be called internally if something different than a string is passed to a node's textContent.

I am aware of that, my point was just that IF there is a use case where somebody passes something to `writeStringTextNode` that overwrites `toString()` and outputs a string that should have been wrapped in CDATA sections as well. Because to support that we would need to call `toString` in `writeStringTextNode` before doing the CDATA checks etc. But if that is not a use case that is perfectly fine for me, i was just asking about.

> In a text node (which we're dealing with here), only "&" and "<" need to be escaped. The regular expression in my code suggestions is already adjusted to only contain these two.

Thanks for confirming that and thanks for providing the regex. I will test this later today and push the changes if everything still looks ok.

> I don't think that's necessary. Instead of using xmleq(), I'd suggest using node.textContent to compare the writeStringTextNode() node with the reference node.

I think using xmleq and adapting the underlying functions makes sense, since this also ensures the behavior is correct for all existing testcases that utilize xmleq (KML, WFS etc).

ahocevar commented:
> I think i traced it back to KML and [this ](https://github.com/openlayers/openlayers/blob/main/test/browser/spec/ol/format/kml.test.js#L1903)testcase, but i am not sure right now. I am not a KML expert so i didn't look into it. The only thing i can say is that i definitly had cases where something not being a string was passed to writeStringTextNode.

So I'd say just leave the `typeof string === 'string'` check there, and call it good.

> I think using xmleq and adapting the underlying functions makes sense, since this also ensures the behavior is correct for all existing testcases that utilize xmleq (KML, WFS etc).

That's a good point. In this case, please also add tests for `xmleq()` in `test/browser/spec/ol/expect.test.js`.

That said, you can leave the tests as you have them now, creating the reference XML with `writeCDATASection()`. It makes it clear what the code is doing, in a beautiful way. No need to change them later when the Chromium issue is fixed.

MelkorCC commented:
Updated PR, changes:
- Replaced `checkCDATA` with suggested regex in `xsd.js`
- Removed unnessesary tests for special XML characters that we don't have to care about when dealing with textnodes in `xsd.test.js`
- Removed test for boolean handling in `xsd.js` (test for number is still in place, i guess it is ok to have one that tests the typeof string handling)
- Replaced `createElementNS` by call to `getDocument().createElement()` in `xsd.test.js`
-  Added new tests in `expect.test.js` for CDATA section handling
- Added some comments in `test-extensions.js` to mark CDATA specific stuff as such

ahocevar reviewed (changes requested):
Looks good, I only added a few minor suggestions.

MelkorCC commented on src/ol/format/xsd.js:
```diff
@@ -147,5 +147,22 @@ export function writeNonNegativeIntegerTextNode(node, nonNegativeInteger) {
-  node.appendChild(getDocument().createTextNode(string));
+  if (
+    typeof string === 'string' &&
+    (/^\s/.test(string) ||
+      /\s$/.test(string) ||
+      /(\n|\t|\r|<|&| {2})/.test(string))
```
I completely agree with that, but funnily enough all regex is in functions in `xsd.js`. Might be something we could consider optimizing in a different PR.

MelkorCC commented on test/browser/spec/ol/format/xsd.test.js:
```diff
@@ -10,4 +15,154 @@ describe('ol/format/xsd', function () {
+      it('can handle basic string data', function () {
+        const text = '<text>Test</text>';
+
+        const node = getDocument().createElement('text');
```
good to know, i will adapt this

MelkorCC commented:
Updated PR:
- added suggested textNode checks for CDATA testcases in `xsd.test.js`
- moved regex outside the function and reuse them

ahocevar reviewed (approved):
Thanks for your effort, @MelkorCC.

MelkorCC commented:
@ahocevar thank you again for your very quick feedback and input.

I have one last question, do you think it is possible to get this into a release before May? We have an important rollout scheduled in May and i would love to have this in it (at the moment i am using a solution that takes the XMLDocument for the WFS transaction and then injects the CDATA sections into it ... far from ideal). I know that there are dev builds, but i would need a "real" release for it to be allowed into the rollout.

ahocevar commented:
@MelkorCC The releases in the `latest` channel run the same tests as the `dev` versions, the only difference is that a human being adds release notes. However, a new `latest` release will likely be published in March or April already - definitely before May.
